Scp image transfer 192

badgerwild opened the pull request:
# Assignee Section

## Description
Created a script to send images using scp. The file name is passed to the script as a command line argument. In order for the script to work, ssh keys need to be setup on both the local and the host computers. This only has to be done once for each computer, before running the script.

Host address, host username, and remote file path all are updated in the config.py file so that this information does not need to be changed in the script itself.  

[If applicable, include a list of main points]
- [ ] Image transfers between computers

closes #[192]

The approval from all software team leads is necessary before merging.

# Reviewer Section

Aside from local testing and the General Integration Test it is implied that static analysis should be included in the verification process.

- [ ] Local Test Performed Successfully
- [ ] [General Integration Test](https://docs.google.com/document/d/1ug0CpA1cIzURP8DDFSvCt2CEJJSwJ6Ta6B1LG_hYk6I/edit) Performed Successfully

For Pull Requests that do not include code changes, it is not required to perform the tests above.

MartensCedric commented on robot/rospackages/src/image_scp/config.py:
```diff
@@ -0,0 +1,6 @@
+#enter user name for host computer
+user = "USER NAME FOR HOST"
+#enter desired image path on host computer
+remote_path_1 = "PATH FOR IMAGE ON HOST COMPUTER"
```
So what happens if we want to dynamically send an image. I want to call this script lets say, everytime I take a new picture. This picture has a new name. Perhaps you want to just put a folder that needs to be synced. But then we will have issues if there's leftover files.

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,45 @@
+"""
+Script takes in the image as a command line argument.
+SSH keys need to be configured on both remote and local machine in order for the script to work.
+Host name, user nmae, and remote path are all imported from a config file. Update the information there 
```
user nmae

MartensCedric commented:
Any progress?

badgerwild commented:
> Any progress?

Yes, I've been working on it this week. Was thinking of the best way to implement your changes. Been trying a few different ways, should have it resolved soon.

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+"""
+Script takes in the image as a command line argument.
+SSH keys need to be configured on both remote and local machine in order for the script to work.
+Host name, user nmae, and remote path are all imported from a config file. Update the information
```
```suggestion
Host name, user name, and remote path are all imported from a config file. Update the information
```

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+Script takes in the image as a command line argument.
+SSH keys need to be configured on both remote and local machine in order for the script to work.
+Host name, user nmae, and remote path are all imported from a config file. Update the information
+there so that this script is portable .
```
```suggestion
there so that this script is portable.
```

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+        self.user = user
+        self.remote_path = remote_path
+        self.file_name = file_name
+#initiate the connection to the host machine
+    def connect(self):
+        self.client = SSHClient()
+        self.client.load_system_host_keys()
+        self.client.connect(hostname=self.host, username=self.user)
+        self.scp = SCPClient(self.client.get_transport())
```
```suggestion
    def connect(self):
        """initiate the connection to the host machine""""
        self.client = SSHClient()
        self.client.load_system_host_keys()
        self.client.connect(hostname=self.host, username=self.user)
        self.scp = SCPClient(self.client.get_transport())
```

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+
+if __name__ == "__main__":
+
+  #takes the image as a command line argument
+    parser = argparse.ArgumentParser(description="the file that gets sent")
```
```suggestion
    #takes the image as a command line argument
    parser = argparse.ArgumentParser(description="the file that gets sent")
```

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+
+    def send_batch_images(self, image_directory):
+        bulk_tranfer = [self.send_image(image) for image in image_directory]
+        print(len(bulk_transfer) + "Images have been transfered")
```
```suggestion
        print(len(bulk_transfer) + " images have been transfered")
```

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+        try:
+            self.scp.put(image, recursive=True, remote_path=self.remote_path)
+        except SCPException as error:
+            raise error
```
```suggestion
            print(error)
```

MartensCedric replied:
You dont have a try catch when you send the image (outside of the function) by catching the error and rethrowing it, nothing useful happens.

MartensCedric replied:
besides the program crashing

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,78 @@
+    image_files = []
+
+    #Takes in a directory name from which a config file and jpeg files are extracted
+    for name in os.listdir('test-file/'):
```
is this `test-file` meant to be used in production or is this a mistake?

badgerwild replied:
No, not meant for production. It's a mistake left over from testing. I'm  addressing your suggestions now.

badgerwild replied:
@MartensCedric  Fixed this issue and debugged the rest of the code. It "Works on my machine" now.

NinjaWeeaboo commented:
Maybe this can be tested on the TX2 for those pro gamer bonus points

MartensCedric commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,79 @@
+import argparse
+from config import (host, user, remote_path_1)
+"""
+Script takes in a directory name as a command line argument. This directory should contain
+a config.txt file and any images that need to be transered.
+SSH keys need to be configured on both remote and local machine in order for the script to work.
+Host name, user name, and remote path are all imported from config.txt file. Update the information
+there so that this script is portable.
```
```suggestion
Script takes in a directory name as a command line argument. This directory should contain
a config.txt file and any images that need to be transfered.
SSH keys need to be configured on both remote and local machines in order for the script to work.
Host name, user name, and remote path are all imported from config.txt file. Update the information
there so that this script is portable.
```

PeterGhimself commented on robot/rospackages/src/image_scp/config.txt:
```diff
@@ -0,0 +1,6 @@
+#Use this as a template for entering user details and destination of files so script doesn't need to be changed.
+#Enter credentials/ path without quotation marks 
+Username replaceWithUse
```
you meant replaceWithUser? I would suggest to make it more clear like `basestationUserName`, though perhaps (since I recommend using a python file) it can be set automatically, by getting the `$USER` environement variable through python.

badgerwild replied:
I added instructions for testing,  do they make sense?

PeterGhimself commented on robot/rospackages/src/image_scp/config.txt:
```diff
@@ -0,0 +1,6 @@
+#Use this as a template for entering user details and destination of files so script doesn't need to be changed.
+#Enter credentials/ path without quotation marks 
+Username replaceWithUse
+Host relaceWithHostDomain
```
while host domain is probably still correct, it's a bit general. This is supposed to be the OBCs IP address right? Maybe we can just put `roverIP` instead?

PeterGhimself commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,79 @@
+import argparse
+from config import (host, user, remote_path_1)
+"""
+Script takes in a directory name as a command line argument. This directory should contain
```
Unforuntately I think this approach isn't the most efficient, it's probably best to make the script target a directory that is remote, the config file can stay local though.

PeterGhimself commented on robot/rospackages/src/image_scp/image_scp.py:
```diff
@@ -0,0 +1,79 @@
+
+"""
+
+class remote_connection:
```
Classes should be PascalCase, refer to the [TLDR](https://namingconvention.org/python/)

PeterGhimself commented:
So I merged master in and added some changes after testing, there is however what I think is a bug/weird behavior.

To reproduce:

1. On host computer make sure you have `/home/[username]/Pictures/` folder with something in it (quickly: `touch yo`).
2. Run `./image_scp.py`, note that a new folder will be created in your home directory called `Picturez` with the files in there
3. Repeat step 2, but notice that it creates a new folder `Picturez` inside of the already existing `Picturez` folder

:thinking: 

Also perhaps we should rename this script? at first our intention was to use this for images but it is in fact a very general purpose script. Could be used to copy over log files, videos, etc. There is no constraint (and I don't think we should impose one) for it being images specifically.

Thoughts?

badgerwild commented:
@PeterGhimself 
I got the behaviour...I'll work on a fix for that.

Renaming the script to a general file transfer makes a lot of sense to me. I'll rename it to file_scp .

badgerwild commented:
> So I merged master in and added some changes after testing, there is however what I think is a bug/weird behavior.
> 
> To reproduce:
> 
>     1. On host computer make sure you have `/home/[username]/Pictures/` folder with something in it (quickly: `touch yo`).
> 
>     2. Run `./image_scp.py`, note that a new folder will be created in your home directory called `Picturez` with the files in there
> 
>     3. Repeat step 2, but notice that it creates a new folder `Picturez` inside of the already existing `Picturez` folder
> 
> 
> thinking
> 
> Also perhaps we should rename this script? at first our intention was to use this for images but it is in fact a very general purpose script. Could be used to copy over log files, videos, etc. There is no constraint (and I don't think we should impose one) for it being images specifically.
> 
> Thoughts?
@PeterGhimself 
I addressed the issue raised here about the unexpected behaviour for file creation.

MartensCedric commented:
Could you provide better steps for testing for I can get some people unfamiliar with the issue to test it. You should explain how to test it locally (what values to put in transferconfig), how to run the scripts, the arguments and the expected output. Then Ill test it and so should @RealWilliamWells and @AFknUsername

badgerwild commented:
@MartensCedric Updated the testing instructions.

MartensCedric commented:
@badgerwild I'm assuming I goofed somewhere 
![image](https://user-images.githubusercontent.com/18620025/99881751-990be500-2be9-11eb-84ca-0b3541cdce7c.png)

Can you specify in the steps the exact ip that you used an all

![image](https://user-images.githubusercontent.com/18620025/99881791-dec8ad80-2be9-11eb-8c57-b5ca87eab1bd.png)

MartensCedric commented:
Just for the future, when you put steps I would suggest to just copy put commands instead. Suppose you didn't edit "transferconfig.py" and someone not familiar with that specific file reviews this PR, he will have to find it. I found it easily because it was an edited file but if it wasn't other reviewers might have trouble finding it, and doing this small hassle for every other PR. That's why I would do steps like this. 

1. vim ~/Programming/robotics-prototype/robot/basestation/file_scp/transferconfig.py
2. Change rover ip to be "encs.concordia.ca"
3. Change rover username to be "(Your ENCS user name)"
...

badgerwild commented:
@MartensCedric  
The domain for concordia should be "login.encs.concordia.ca" Also, it looks like I goofed and didn't commit the updated config file, so it wouldn't have worked anyways. I'll update the steps following your recommendations .

ginywiny reviewed (commented):
Upon running with the new config, I ended up with the following issue. 
![image](https://user-images.githubusercontent.com/36121273/99885817-915a3980-2c05-11eb-9d36-e777b3d76423.png)

Has this ever happened to you @MartensCedric 

I have followed the new instructions, and set `rover_ip` to 'login.encs.concordia.ca', `rover_user` to my encs username 'm_lesk', I created a directory under my encs account called 'SpaceConcordia' and set `rover_image_dir` to '~/SpaceConcordia', lastly I set `payload_file_subdir` to 'TestFile', while I left `key_path` and `rsa_password` untouched.

badgerwild commented:
@ginywiny 
I'll try and have a look at that later today and see if I get issues.

AFknUsername commented:
I get the same error as @ginywiny when using the same settings
I also tried setting a custom ssh key to see if it would make a difference, and it does not work. I get an error :`There is a problem with the key, double check that your encryption password is correct` 
The key is not encrypted.

badgerwild commented:
> I get the same error as @ginywiny when using the same settings
> I also tried setting a custom ssh key to see if it would make a difference, and it does not work. I get an error :`There is a problem with the key, double check that your encryption password is correct`
> The key is not encrypted.

I'll try and figure out what is happening.

badgerwild commented:
@ginywiny @VimFreeman Following the stand up today, I changed the ssh key from rsa to ed25519 and that seemed to fix the problems we were having. The ed25519 also seems to produce a significantly faster runtime for the script when compared to the rsa key.

VimFreeman commented:
Closing since we do not need it anymore.
